### PR TITLE
mitigate ABA with 32-bit queue indices if possible

### PR DESCRIPTION
**Pull request by sbarral**
When 64-bit atomics are supported, use 32-bit queue indices. This greatly improves resilience to ABA and has no impact on performance on 64-bit platforms.

Fixes: #5041

## Motivation

Currently, a stealer preempted at the wrong moment could attempt to steal more tasks than are available in the queue, resulting in UB. This is due to a potential ABA where the stealer head wraps around due to many push/pop cycles, leading the stealer to wrongly conclude that the state of the queue has not changed.

## Solution

The proposed solution is to strongly improve ABA resilience by using 32-bit instead of 16-bit buffer indices when possible.

At the moment, "when possible" is determined based on the availability of 64-bit atomics, as reported by the `cfg_has_atomic_u64` macro. Another solution would be to use `cfg(target_pointer_width = "64")` to exclude 32-bit targets that support 64-bit atomics natively, such as `armv7`, since these targets may see an impact in performance.

No performance regression was observed on local tests on a `x86_64` target using the `rt_multi_threaded` benchmark.

### CODE REVIEW

File: tokio/src/runtime/scheduler/multi_thread/queue.rs [hunk 1]
```diff
@@ -10,6 +9,20 @@ use std::mem::MaybeUninit;
 use std::ptr;
 use std::sync::atomic::Ordering::{AcqRel, Acquire, Relaxed, Release};
 
+// Use wider integers when possible to increase ABA resilience.
```
Comment (Darksonn) [inline tokio/src/runtime/scheduler/multi_thread/queue.rs:12]: We should probably include a link to  #5041 here.

File: tokio/src/runtime/scheduler/multi_thread/queue.rs [hunk 2]
```diff
@@ -10,6 +9,20 @@ use std::mem::MaybeUninit;
...
+    pub(crate) type UnsignedLong = u64;
+    pub(crate) type AtomicUnsignedShort = crate::loom::sync::atomic::AtomicU32;
+    pub(crate) type AtomicUnsignedLong = crate::loom::sync::atomic::AtomicU64;
+}
+cfg_not_has_atomic_u64! {
+    pub(crate) type UnsignedShort = u16;
+    pub(crate) type UnsignedLong = u32;
+    pub(crate) type AtomicUnsignedShort = crate::loom::sync::atomic::AtomicU16;
+    pub(crate) type AtomicUnsignedLong = crate::loom::sync::atomic::AtomicU32;
+}
```
Comment (Darksonn) [inline tokio/src/runtime/scheduler/multi_thread/queue.rs:24]: Seems like these don't need `pub(crate)`.

**Review by Darksonn — COMMENTED** [review]
In general this looks fine.

File: tokio/src/runtime/scheduler/multi_thread/queue.rs (hunk 2 above)
Comment (sbarral) [reply tokio/src/runtime/scheduler/multi_thread/queue.rs:24]: oops, sorry, a bad case of copy-pasta...

**Review by Darksonn — APPROVED** [review]

**Comment by Darksonn** [comment]
Please merge master into your branch to fix the CI issues.

**Comment by Darksonn** [comment]
Nevermind, I could do it myself.

File: tokio/src/runtime/scheduler/multi_thread/queue.rs [hunk 3]
```diff
@@ -10,6 +9,22 @@ use std::mem::MaybeUninit;
...
+    type UnsignedLong = u64;
+    type AtomicUnsignedShort = crate::loom::sync::atomic::AtomicU32;
+    type AtomicUnsignedLong = crate::loom::sync::atomic::AtomicU64;
+}
+cfg_not_has_atomic_u64! {
+    type UnsignedShort = u16;
+    type UnsignedLong = u32;
+    type AtomicUnsignedShort = crate::loom::sync::atomic::AtomicU16;
+    type AtomicUnsignedLong = crate::loom::sync::atomic::AtomicU32;
+}
```
Comment (Darksonn) [inline tokio/src/runtime/scheduler/multi_thread/queue.rs:26]: Unfortunately, the `BITS` constant is not available on the minimum Rust version we support.
```
error[E0658]: use of unstable library feature 'int_bits_const'
   --> tokio/src/runtime/scheduler/multi_thread/queue.rs:526:58
    |
526 |     (real as UnsignedLong) | ((steal as UnsignedLong) << UnsignedShort::BITS)
    |                                                          ^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #76904 <https://github.com/rust-lang/rust/issues/76904> for more information
```
You will need to define the constant yourself:
```suggestion
cfg_has_atomic_u64! {
    type UnsignedShort = u32;
    const SHORT_BITS: u32 = 32;
    type UnsignedLong = u64;
    type AtomicUnsignedShort = crate::loom::sync::atomic::AtomicU32;
    type AtomicUnsignedLong = crate::loom::sync::atomic::AtomicU64;
}
cfg_not_has_atomic_u64! {
    type UnsignedShort = u16;
    const SHORT_BITS: u32 = 16;
    type UnsignedLong = u32;
    type AtomicUnsignedShort = crate::loom::sync::atomic::AtomicU16;
    type AtomicUnsignedLong = crate::loom::sync::atomic::AtomicU32;
}
```

**Review by Darksonn — APPROVED** [review]
Thanks.

**Review by Noah-Kennedy — APPROVED** [review]

File: tokio/src/runtime/scheduler/multi_thread/queue.rs (hunk 3 above)
Comment (Darksonn) [reply tokio/src/runtime/scheduler/multi_thread/queue.rs:26]: I'm fine either way. What you did is also fine.

**Comment by Darksonn** [comment]
You don't need to add any more merge commits at this point.

**Pull request merged**